gradio/queue.py

```diff
@@ -69,6 +85,26 @@ def get_active_worker_count(self) -> int:
                 count += 1
         return count
 
+    def get_events_in_batch(self) -> Tuple[List[Event] | None, bool]:
+        if not (self.event_queue):
+            return None, False
+
+        first_event = self.event_queue.pop(0)
+        events = [first_event]
+
+        event_fn_index = first_event.fn_index
+        batch = self.blocks_dependencies[event_fn_index]["batch"]
+
+        if batch:
+            batch_size = self.blocks_dependencies[event_fn_index]["max_batch_size"]
+            rest_of_batch = [
+                event for event in self.event_queue if event.fn_index == event_fn_index
+            ][: batch_size - 1]
+            events.extend(rest_of_batch)
```

Comment:
I strongly think that this should not be handled like this, there are 2 advantages 
1. This will search the whole queue, not efficient
2. It will not respect the queue rank for other event types

I suggest this approach, keep a batch_list for every batched event, and save batchy events
```

batch_wait_list = []
...


while loop
...
event = queue.pop()

if event.batch:
  batch_wait_list.append(event)
  if batch_wait_duration > time_threshold:
    await process_batch(batch_wait_list);
  if len(batch_wait_list) < batch_size:
    continue
  else:
    await process_batch(batch_wait_list);

else:
  await process_event(event)
```

Reply:
Thanks @FarukOzderim for the feedback, what do you mean by "It will not respect the queue rank for other event types"?

Reply:
well, when the batched event gets popped, it finds the other several batched events irrespective of their rank. Even if they are in the end of the queue.

Reply:
@FarukOzderim I may be misunderstanding, but the approach you are proposing will not work because it assigns all batched events to a single `batch_wait_list`. But you can only process events in a batch if they share the same underlying function (`fn_index`). 

If you are suggesting that we have separate `batch_wait_lists` for each event, then it effectively becomes the same approach as what I have already implemented, except that it is unclear to me in your implementation how to select between different `batch_wait_lists` when a worker is free.

> well, when the batched event gets popped, it finds the other several batched events irrespective of their rank. Even if they are in the end of the queue.

This is not accurate, as what happens is `rest_of_batch` takes the _first_ `max_batch_size` events in the event queue that share the same function index. So it does respect the order of the queue, but it prioritizes events that share the same function index in order to facilitate batching.

Reply:
> well, when the batched event gets popped, it finds the other several batched events irrespective of their rank. Even if they are in the end of the queue.
This is irrespective of their rank compared to unbatched events.
Extreme example, assume there are 2 batched events in the queue,
1st one is in the 1st rank
2nd one is in the 1000st rank
In this case rank of events in 2-999 are not respected.

> If you are suggesting that we have separate batch_wait_lists for each event, then it effectively becomes the same approach as what I have already implemented, except that it is unclear to me in your implementation how to select between different batch_wait_lists when a worker is free.

It is not the same thing, first of all with your approach there is a O(n) list copy-search everytime a batched event is popped, which will hurt the performance considerably. The difference lies in performance and execution time. Current version executes with the first batched event, my suggestion executes with the last batch event. (considering the batch-size) 
Selecting a batch_list is quite trivial actually, let me make the pseudo code more detailed.

```

batch_wait_list_list = [ [] * fn_index_count ]

...


while loop
...
event = queue.pop()

if event.batch:
  batch_wait_list = batch_wait_list_list[event.fn_index]
  batch_wait_list.append(event)
  if batch_wait_duration > time_threshold: #Never mind threshold for now, it can also be tracked with a list
    await process_batch(batch_wait_list);
  if len(batch_wait_list) < batch_size:
    continue
  else:
    await process_batch(batch_wait_list);

else:
  await process_event(event)
```

Reply:
Ah that's much clearer now @FarukOzderim. However, this approach has one problem in my mind, which is that it could result in an arbitrarily long execution time for a user whose request is at the beginning of the queue, if their batch never "fills up". Imagine the following queue (I'm just writing the `fn_index` for each event), in which event `1` is a batch event:

`[1, 2, 2, 2, 2, 2 ...]`

If we get an influx of event `2`s, event `1` will never get processed. That is why I prefer the approach of going from the beginning of the queue, because it will never result in arbitrarily long wait times for people at the beginning of the queue. If a user is in position $i$, and the runtime for a batch is $t$, we can guarantee that the user has to wait less than $i \cdot t$.

Now in theory this could overcome with an event-specific `batch_wait_duration`, but I would really like to avoid adding that hyperparameter at this time, because if it's improperly set, it could result in a *worse* performance with batching than without.

That being said, I agree with your point about the look-up time being $O(n)$ and how that could be expensive if the queue size is large. Let me benchmark it and if it is expensive, then I'll use a different structure that will allow $O(1)$ lookups.

Reply:
@abidlabs IMO the solution to that problem is `batch_wait_duration_threshold` and compromising on other parts is not a good idea. I don't think setting this threshold as 60 secs is a problem. If batch is being used than letting people wait is just expected.

Nonetheless if you are too determined about this just start a search from the beginning of the queue, instead of copying and searching all the queue over here
```
rest_of_batch = [
                event for event in self.event_queue if event.fn_index == event_fn_index
            ][: batch_size - 1]
```

Reply:
@FarukOzderim I profiled this code and the latency is not a big concern. Sample code:

```py
import time
import random

queue = [{'index': random.randint(0, 5)} for _ in range(10000)]

t = time.time()
rest_of_batch = [e for e in queue if e['index'] == 3][:30]
print(time.time() - t)
```

You can see that the latency is ~1 ms even for a queue size of 10k items.

Reply:
Calling `remove` may be inefficient cause I think python has to move all the elements of the list whenever you remove something but I suspect it would only be noticed for very large batches (which are rare)

Reply:
@abidlabs I think you don't take into remove overhead into account.

Here is the time comparison between these two versions, there is ~50x speedup in my runs.
This would unnecessarily slow down demos with short inference times significantly. But would not affect long inference times that much. Since there is no loss but some gain in the second version I recommend going with V2.

Time of the V1: 0.0877530574798584
Time of the V2: 0.0018298625946044922
Speedup: 47.95609120521173


```

import time

queue = [1, 2, 3, 4, 5] * 1000

start = time.time()

while queue:
    element = queue.pop()
    rest_of_batch = [e for e in queue if e == element][:5]
    [queue.remove(e) for e in rest_of_batch]

end = time.time()
v1_time = end-start

print(f"Time of the V1: {v1_time}")

queue = [1, 2, 3, 4, 5] * 1000
batch_wait_list = [[] for i in range(5)]

start = time.time()

while queue:
    element = queue.pop()
    batch_wait_list[element-1].append(element)
    if len(batch_wait_list[element-1]) == 5:
        batch_wait_list[element-1] = []
end = time.time()
v2_time = end-start
print(f"Time of the V2: {v2_time}")
print(f"Speedup: {v1_time/v2_time}")

```

Reply:
Hey @FarukOzderim thanks for providing this profiling code. Agreed I misunderstood the part about `.remove()` taking an unnecessary amount of time. I still feel like the absolute difference is not great (we're talking about 80 ms to clear a queue of 5,000 events) and I'd like to get this PR in without significant additional refactoring, so let me merge this in as is, but I agree the way you propose would be better, so I'll revisit in a future PR (or feel free to make a PR and I'll review & merge)